### PR TITLE
Introduce DIExpressionOptimizer

### PR DESCRIPTION
**Pull request by rastogishubham**
DIExpressionOptimizer is a lightweight class that can be used to optimize DIExpressions that can get very large, by involving some simple constant folding and removing appends of unnecessary operations.

For example:

If we have a `DIExpression: {dwarf::DW_OP_LLVM_arg, 0, dwarf::DW_OP_constu, 2, dwarf::DW_OP_plus}` and want to append `{dwarf::DW_OP_constu, 3, dwarf::DW_OP_plus}`

Instead of having the final `DIExpression be {dwarf::DW_OP_LLVM_arg, 0, dwarf::DW_OP_constu, 2, dwarf::DW_OP_plus, dwarf::DW_OP_constu, 3, dwarf::DW_OP_plus}. `

It will be reduced to `{dwarf::DW_OP_LLVM_arg, 0, dwarf::DW_OP_constu, 5, dwarf::DW_OP_plus}`

There is an implementation for `DW_OP_mul` and `DW_OP_plus_uconst` as well

It also just flat out does not append `{DW_OP_constu, 1, DW_OP_mul}, {DW_OP_constu, 1, DW_OP_div}, {DW_OP_constu, 0, DW_OP_plus}, {DW_OP_constu, 0, DW_OP_minus}, {DW_OP_constu, 0, DW_OP_shl}, {DW_OP_constu, 0, DW_OP_shr}`

Hopefully this can be improved upon by the larger community to add more optimizations in the future.

This patch is part of a stacked diff, I was not sure how to do this any better because github squashes multiple commits in one PR, but the PR this depends on is:

https://github.com/llvm/llvm-project/pull/69768

### CODE REVIEW

**Comment by pogo59** [comment]
how does this relate to (or supersede) DIExpression::constantFold() ?

**Comment by pogo59** [comment]
I suspect we would prefer not to have two separate constant folders. Can you combine them?

File: llvm/include/llvm/IR/DebugInfoMetadata.h [hunk 1]
```diff
@@ -3073,6 +3073,109 @@ template <> struct DenseMapInfo<DIExpression::FragmentInfo> {
   static bool isEqual(const FragInfo &A, const FragInfo &B) { return A == B; }
 };
 
+class DIExpressionOptimizer {
```
Comment (adrian-prantl) [inline llvm/include/llvm/IR/DebugInfoMetadata.h:3076]: Can you add a Doxygen comment for the entire class with just a sentence on what this does?

File: llvm/lib/IR/DebugInfoMetadata.cpp [hunk 2]
```diff
@@ -1734,6 +1734,143 @@ const DIExpression *DIExpression::extractAddressClass(const DIExpression *Expr,
...
+    auto Size = Cursor.peek()->getSize();
+    Loc += Size;
+    Cursor.consume(1);
+  }
+
+  return OpLocs;
+}
+
+bool DIExpressionOptimizer::operatorIsCommutative(uint64_t Operator) {
+  return Operator == dwarf::DW_OP_mul || Operator == dwarf::DW_OP_plus;
```
Comment (adrian-prantl) [inline llvm/lib/IR/DebugInfoMetadata.cpp:1754]: How about DW_OP_and, or, xor?

File: llvm/lib/IR/DebugInfoMetadata.cpp [hunk 3]
```diff
@@ -1734,6 +1734,143 @@ const DIExpression *DIExpression::extractAddressClass(const DIExpression *Expr,
...
+  case dwarf::DW_OP_div:
+  case dwarf::DW_OP_shl:
+  case dwarf::DW_OP_shr:
+    return true;
+  default:
+    return false;
+  }
+}
+
+void DIExpressionOptimizer::reset() {
```
Comment (adrian-prantl) [inline llvm/lib/IR/DebugInfoMetadata.cpp:1772]: Is this needed for performance reasons? If not, it is easier to reason about the code if a fresh DIExpressionOptimizer is created for each use.

File: llvm/lib/IR/DebugInfoMetadata.cpp [hunk 4]
```diff
@@ -1734,6 +1734,143 @@ const DIExpression *DIExpression::extractAddressClass(const DIExpression *Expr,
...
+  NewMathOperator = 0;
+  CurrMathOperator = 0;
+  OperatorLeft = 0;
+  OperatorRight = 0;
+}
+
+void DIExpressionOptimizer::optimize(SmallVectorImpl<uint64_t> &NewOps,
+                                     ArrayRef<uint64_t> Ops) {
+
+  if (Ops.size() == 2 && Ops[0] == dwarf::DW_OP_plus_uconst) {
```
Comment (adrian-prantl) [inline llvm/lib/IR/DebugInfoMetadata.cpp:1782]: Why == 2 and not >= 2?

File: llvm/include/llvm/IR/DebugInfoMetadata.h [hunk 5]
```diff
@@ -3073,6 +3073,109 @@ template <> struct DenseMapInfo<DIExpression::FragmentInfo> {
...
+  // NewMathOperator = DW_OP_minus
+  // OperandRight = 3
+  // OperatorRight = DW_OP_consts
+  // CurrMathOperator = DW_OP_plus
+  // OperandLeft = 2
+  // OperandLeft = DW_OP_constu
+
+  /// The math operator of the new subexpression being appended to the
+  /// DIExpression.
+  uint64_t NewMathOperator = 0;
```
Comment (adrian-prantl) [inline llvm/include/llvm/IR/DebugInfoMetadata.h:3089]: What do you think about calling `MathOperator` more generically `BinOp`, or `BinaryOperator`, so all operations that take two arguments fit the description?

File: llvm/lib/IR/DebugInfoMetadata.cpp [hunk 6]
```diff
@@ -1734,6 +1734,143 @@ const DIExpression *DIExpression::extractAddressClass(const DIExpression *Expr,
...
+    // Currently, DIExpressionOptimizer only supports arithmetic operations
+    // such as Ops = {DW_OP_constu, 1, DW_OP_mul}
+    NewOps.append(Ops.begin(), Ops.end());
+    return;
+  }
+
+  if (OperatorRight != dwarf::DW_OP_constu &&
+      OperatorRight != dwarf::DW_OP_consts) {
+    NewOps.append(Ops.begin(), Ops.end());
+    return;
```
Comment (adrian-prantl) [inline llvm/lib/IR/DebugInfoMetadata.cpp:1802]: Why is the state preserved in this case but not in all the others?

**Review by adrian-prantl — COMMENTED** [review]
I think in general this looks very useful. 
Can you explain why DIExpressionOptimizer needs to keep state?

**Comment by adrian-prantl** [comment]
> I suspect we would prefer not to have two separate constant folders. Can you combine them?

+1, that sounds like a good idea!

**Pull request closed without merging**